support/k8s/k8s-workload-registrar/controllers/pod_controller.go

```diff
@@ -0,0 +1,221 @@
+		}
+
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	spiffeIdUri := ""
+	switch r.Mode {
+	case PodReconcilerModeServiceAccount:
+		spiffeIdUri = r.makeID("ns/%s/sa/%s", req.Namespace, pod.Spec.ServiceAccountName)
+	case PodReconcilerModeLabel:
+		if val, ok := pod.GetLabels()[r.Value]; ok {
+			spiffeIdUri = r.makeID("%s", val)
+		} else {
+			// No relevant label
+			return ctrl.Result{}, nil
+		}
+	case PodReconcilerModeAnnotation:
+		if val, ok := pod.GetAnnotations()[r.Value]; ok {
+			spiffeIdUri = r.makeID("%s", val)
+		} else {
+			// No relevant annotation
+			return ctrl.Result{}, nil
+		}
+	}
+
+	spiffeId := &spiffeidv1beta1.SpiffeID{
+		ObjectMeta: v1.ObjectMeta{
+			Namespace:   pod.Namespace,
+			Annotations: make(map[string]string),
+		},
+		Spec: spiffeidv1beta1.SpiffeIDSpec{
+			SpiffeId: spiffeIdUri,
+			DnsNames: make([]string, 0),
+			Selector: spiffeidv1beta1.Selector{
+				PodUid:    pod.GetUID(),
+				Namespace: pod.Namespace,
+			},
+		},
+	}
+
+	// Set pod as owner of new SPIFFE ID
+	err := controllerutil.SetControllerReference(&pod, spiffeId, r.Scheme)
+	if err != nil {
+		r.c.Log.WithFields(logrus.Fields{
+			"SpiffeID.Name": spiffeId.Name,
+		}).WithError(err).Error("Failed to set pod as owner of new SpiffeID")
+		return ctrl.Result{}, err
+	}
+
+	// Set pod name as first DNS name
+	spiffeId.Spec.DnsNames = append(spiffeId.Spec.DnsNames, pod.Name)
+
+	err = r.createSpiffeId(ctx, pod.ObjectMeta.Name, spiffeId)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Add label to pod with name of SPIFFE ID
+	if pod.ObjectMeta.Labels["spiffe.io/spiffeid"] != spiffeId.ObjectMeta.Name {
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of Pod before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
+				r.c.Log.WithError(err).Error("Failed to get latest version of Pod")
+				return err
+			}
+			pod.ObjectMeta.Labels["spiffe.io/spiffeid"] = spiffeId.ObjectMeta.Name
+
+			err = r.Update(ctx, &pod)
+
+			return err
+		})
+		if retryErr != nil {
+			r.c.Log.WithError(retryErr).Error("Update failed")
+			return ctrl.Result{}, retryErr
+		}
+		r.c.Log.Info("Added label to pod")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// createSpiffeId attempts to create the SPIFFE ID object. It uses the pod name appeneded with hash of the SPIFFE ID Spec
+// to create a unique name.
+func (r *PodReconciler) createSpiffeId(ctx context.Context, podName string, spiffeId *spiffeidv1beta1.SpiffeID) error {
+	var collisionCount int32
+	var existing spiffeidv1beta1.SpiffeID
+	for {
+		spiffeId.ObjectMeta.Name = podName + "-" + computeHash(&spiffeId.Spec, &collisionCount)
```

Comment:
Would using the pod UID itself instead of a hash work?

Reply:
Yea, and it would things a lot simpler, but its not human readable looking through a list of spiffeids. Using the hash the output is easier to read and determine what the spiffeid resource is for:
```
$ kubectl get spiffeids
NAME                                         AGE
details-v1-74f858558f-j5vrj-6744667c5c       5h35m
productpage-v1-8554d58bff-jgclf-679c9f954d   5h35m
ratings-v1-7855f5bcb9-hllkv-59f494bbd        5h35m
reviews-v1-59fd8b965b-t4zst-69f9b57fbf       5h35m
reviews-v2-d6cfdb7d6-566xd-86fbc9765         5h35m
reviews-v3-75699b5cfb-gbzz7-646fb55cd4       5h35m
```

Reply:
Each pod gets a single SpiffeID CRD, correct? Isn't the pod name in the object meta (i.e. `podName`, e.g., `details-v1-74f858558f-j5vrj`), already unique? I don't understand what the collision bit is for.

Reply:
While pod names are unique they can be rapidly reused (the obvious case being stateful sets), so the reconciliation process can end up with two SpiffeId resources existing for the same pod name (but different resource UIDs.)

Reply:
Yea, thats the reason for the extra hash